Hierarchical state machines

bergundy opened the pull request:
## Notes for reviewers

- I don't consider this a final approach but I do think it's a step in the right direction, we need to model more state machines on top of this to form a more solid API
- This PR is part one of two or maybe three for this refactoring work
- This PR **does not compile**, I cherry picked the `hsm` and `plugins` directories from the WIP sub-state-machines branch
- I'll open another PR that compiles with the "backend" for this framework shortly
- I closed the original PR, which has some review comments #5446

## What changed?

- Modified the statemachines abstraction to be a bit more generic
- Rewrote callbacks as a plugin using this framework

## Why?

This centralizes most the callback code in the plugin directory instead of having it spread out the entire project moving common concerns such as staleness checks, task generation, and (in the future) replication into a framework and should generally help speed up feature delivery and maintainability.

I plan to leverage this framework when implementing Nexus operations.

## How did you test it?

Existing tests from the feature branch and added unit tests.

yycptt commented on service/history/hsm/tree.go:
```diff
@@ -0,0 +1,366 @@
+
+// Node gets an [Node] for a given state machine ID.
+func (c Collection[T]) Node(stateMachineID string) (*Node, error) {
+	return c.node.Child([]Key{{Type: c.Type, ID: stateMachineID}})
```
so a collection contains all state machines of a certain type that are the **_immediate_** children of a given node?

bergundy replied:
> // A Collection of similarly typed sibling state machines.

Does that answer your question?

yycptt reviewed (commented):
I just skimmed through the callback part. They are similar to the previous impl?

tdeebswihart commented on plugins/callbacks/executors.go:
```diff
@@ -0,0 +1,154 @@
+
+	"github.com/nexus-rpc/sdk-go/nexus"
+	commonpb "go.temporal.io/api/common/v1"
+	"go.temporal.io/server/service/history/hsm"
```
I wish we'd gone with the standard `fsm` name here. HSM keeps making me think of Hardware Security Modules rather than state machines

dandavison replied:
My vote is for the standard `fsm` -- we won't forget they're hierarchical.

tdeebswihart commented on plugins/callbacks/executors.go:
```diff
@@ -0,0 +1,154 @@
+func (e activeExecutor) loadUrlAndCallback(ctx context.Context, env hsm.Environment, ref hsm.Ref) (string, nexus.OperationCompletion, error) {
+	var url string
+	var completion nexus.OperationCompletion
+	err := env.Access(ctx, ref, hsm.AccessRead, func(node *hsm.Node) error {
```
What prevents a read access from mutating the hsm.Node struct, anything?

tdeebswihart replied:
We discussed this elsewhere and there's not really a performant solution in Go. We could:
- copy the data and provide it that the accessor
- hash the data before and after and compare to detect mutations

but those both have overhead. Until we discover the mythical Go borrow-checker (😞) there isn't a good and high-performance solution to this.

bergundy replied:
I think you have a good point though, and it might be better to give you a fresh copy of the node data instead of the cached copy or find a cheaper mechanism to ensure safe access.

tdeebswihart replied:
If we rehydrated the data blob on every access we wouldn't have this issue. How often are we actually accessing an individual HSM?

bergundy replied:
I'd say fairly often.

Let's merge this as is and continue discussion and iterate.

tdeebswihart commented on plugins/callbacks/statemachine.go:
```diff
@@ -0,0 +1,229 @@
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// Unique type identifier for this state machine.
+var StateMachineType = hsm.MachineType{
+	ID:   2,
+	Name: "callbacks.Callback",
+}
```
This is a globally unique identifier, right? If so, do we have any static or CI-level enforcement that this is actually unique across all machines?

tdeebswihart commented on proto/internal/temporal/server/api/persistence/v1/executions.proto:
```diff
@@ -394,14 +390,23 @@ message Checksum {
-    // ID of this callback, used as a key in the mutable state map.
-    string id = 3;
-    temporal.api.workflow.v1.CallbackInfo public_info = 4;
+    // Maximum state transitions count for the given version.
+    int64 max_transition_count = 2;
```
Is this the _maximum_ state transition perceived during the specified namespace_failover_version? The terminology here is confusing

bergundy replied:
Yes. I'll try to clarify the comment.

bergundy replied:
Done.

tdeebswihart commented on proto/internal/temporal/server/api/persistence/v1/hsm.proto:
```diff
@@ -0,0 +1,74 @@
+
+// A node in a hierarchical state machine tree.
+message StateMachineNode {
+    // Number of transitions on this state machine object.
+    // Used to verify that a task is not stale if the state machine does not allow concurrent task execution.
+    int64 transition_count = 1;
```
Do we currently allow concurrent operations on a machine and if so how are we verifying those tasks aren't stale?

bergundy replied:
We may allow it. Right now I don't see a requirement for it but I mention in in the docstring for `Ref` that if the `machine_transition_count` is zero that field is ignored and it would be up to the task executor to validate that the task hasn't gone stale.

tdeebswihart commented on proto/internal/temporal/server/api/persistence/v1/hsm.proto:
```diff
@@ -0,0 +1,74 @@
+    // Number of transitions on the corresponding mutable state object. Used to verify that a task is not referencing a
+    // stale state or, in some situations, that the task itself is not stale.
+    int64 mutable_state_transition_count = 3;
+    // If non-zero, the state machine is assumed to support only non-concurrent tasks, and this number should match the
+    // number of state transitions on the corresponding state machine object.
+    int64 machine_transition_count = 4;
```
Is this the _current_ number of transitions executed at the time this Ref is instantiated?

bergundy replied:
Yes.
I can add a comment.

tdeebswihart commented on proto/internal/temporal/server/api/persistence/v1/hsm.proto:
```diff
@@ -0,0 +1,74 @@
+    string id = 2;
+}
+
+message StateMachineRef {
```
A StateMachineRef is a pointer to a state machine _at a certain point in time_, is that correct? Where we're measuring state machine "time" using the `(failover_version, transition_count)` tuple

bergundy replied:
Yes. I'll document this.

tdeebswihart commented on service/history/hsm/tree.go:
```diff
@@ -0,0 +1,366 @@
+	// Parent node. Nil if current node is the root.
+	Parent      *Node
+	registry    *Registry
+	cache       *cachedMachine
+	persistence *persistencespb.StateMachineNode
```
Why are we keeping both the live and serialized version of a machine in memory at once? Is there really a benefit from serializing/deserializing them lazily at the cost of ~2x memory usage (probably more)?

bergundy replied:
That's a good point.
I'm doing this so I can revert back to the old state when a transition mutates a state and fails.
A user may want handle that error and retry a different transition, leaving the state consistent.

tdeebswihart replied:
Yeah, I noticed that when I thought through how rollback works. The only reason our rollback works is because we discard the in-memory state; if we didn't we've nothing to prevent the access fn from both mutating the state _and_ returning an error, corrupting the state machine.

For rollback to work we need a pre-image to switch back to anyways, and we can either keep the serialized persistence version _or_ make a dynamic copy before running the transition function.

tdeebswihart commented on service/history/hsm/tree.go:
```diff
@@ -0,0 +1,366 @@
+	if err != nil {
+		return err
+	}
+	n.persistence.Data = serialized
```
It doesn't look like it

bergundy replied:
Yes @yycptt. @tdeebswihart what makes you think otherwise?

tdeebswihart replied:
If two calls to `MachineTransition` occur within a single transaction I don't see anywhere we handle rolling back the result of prior calls.

Or am I misunderstanding what a transaction is in this context?

tdeebswihart replied:
Ah I see, we're relying on the caller to call `ClearTransactionState()` in case the transaction fails. We don't _really_ have a transaction abstraction or even documentation on what a transaction is in the context of this library and we really need one.

yycptt replied:
I think I understand now. The impl here always you to retry a failed transition from the last succeed transition. If the entire transaction failed, we will clear the entire in memory state. That's not done by the ClearTransactionState() I think, that one only clears the transition output, now the serialized state. It's (will be) done by implementation of the `Access` method which is not in this PR. Basically `Access` is the one providing the transaction abstraction. Is that correct @bergundy or I am still missing something ? 

One more question, who is responsible for calling `ClearTransactionState()`? Also the impl of `Access` I assume ?

bergundy replied:
Yes, all of this is in the next PR.

bergundy commented on proto/internal/temporal/server/api/persistence/v1/hsm.proto:
```diff
@@ -0,0 +1,74 @@
+message StateMachineRef {
+    // Nested path to a state machine.
+    repeated StateMachineKey path = 1;
+    // Namespace failover version on the corresponding mutable state object, used for staleness detection when global
+    // namespaces are enabled.
+    int64 mutable_state_namespace_failover_version = 2;
+    // Number of transitions on the corresponding mutable state object. Used to verify that a task is not referencing a
+    // stale state or, in some situations, that the task itself is not stale.
+    int64 mutable_state_transition_count = 3;
```
Yes.

bergundy replied:
On second thought I'll keep it as is. I don't want another message definition just for this and `VersionedTransition` has a `max_transition_count` where the field name I'd want here is `transition_count`.
Also the docstring would become much more complex for `VersionedTransition`.

bergundy commented on service/history/hsm/registry.go:
```diff
@@ -0,0 +1,127 @@
+
+// RegisterExecutor registers an [Executor] for the given task types.
+// Returns an [ErrDuplicateRegistration] if an executor for any of the types has already been registered.
+func RegisterExecutor[T Task](r *Registry, t int32, exec Executor[T]) error {
```
Yes. I wish Go allowed generic methods...

bergundy commented on service/history/hsm/tree.go:
```diff
@@ -0,0 +1,366 @@
+	data any
+	// Cached children.
+	children map[Key]*Node
+	outputs  []TransitionOutput
```
Yes.

bergundy commented on service/history/hsm/tree.go:
```diff
@@ -0,0 +1,366 @@
+		persistence: machine,
+	}
+	n.cache.children[key] = child
+	return child, nil
```
Ouch, thanks for pointing this out!

bergundy reviewed (commented):
Thanks for the comments.

I'll address ASAP.

bergundy commented:
It's a bit tricky for me to address all of the comments in this PR because it doesn't compile as a standalone thing.

I'll address in the next PR.

bergundy commented:
Addressed comments here: https://github.com/temporalio/temporal/pull/5495/commits/3fd5248569b965a14c8c95aa623e3d52470cafd7

Can we approve this first PR?

yycptt commented on proto/internal/temporal/server/api/persistence/v1/executions.proto:
```diff
@@ -161,8 +162,11 @@ message WorkflowExecutionInfo {
     map<string, temporal.server.api.update.v1.UpdateInfo> update_infos = 79;
-    // Map of callback ID to callback info.
-    map<string, CallbackInfo> callbacks = 80;
+    repeated VersionedTransition transition_history = 80;
```
https://github.com/temporalio/temporal/pull/5495#discussion_r1517503199

yycptt commented on proto/internal/temporal/server/api/persistence/v1/executions.proto:
```diff
@@ -277,22 +276,19 @@ message ArchivalTaskInfo {
 
-// serialized callback task
-message CallbackTaskInfo {
+message OutboundTaskInfo {
```
https://github.com/temporalio/temporal/pull/5495#discussion_r1517370589

bergundy commented:
Merged part 1 into this PR.
I'll make sure to edit the commit message to reflect pre-squash.
